### ISSUE TITLE
Wrap around of input text length in AES-GCM

### ISSUE THREAD

**Issue by amatyuko-intc** [openssl/openssl#17220]
According to the NIST Special Publication 800-38D (Recommendation for GCM mode, p.5.2.1.1 Input Data) the length of input text shall be limited by `2^39-256` bits.

And this check is implemented in `CRYPTO_gcm128_encrypt_ctr32` and related functions.

But when it comes to AES-NI ASM enabled provider's implementation (in function `generic_aes_gcm_cipher_update`), the check is not performed and it allows wrap around of the overall input message length.

```c
static int generic_aes_gcm_cipher_update(PROV_GCM_CTX *ctx, const unsigned char *in,
                                         size_t len, unsigned char *out)
{
    if (ctx->enc) {
        if (ctx->ctr != NULL) {
#if defined(AES_GCM_ASM)
            size_t bulk = 0;

            if (len >= AES_GCM_ENC_BYTES && AES_GCM_ASM(ctx)) {
                size_t res = (16 - ctx->gcm.mres) % 16;

                if (CRYPTO_gcm128_encrypt(&ctx->gcm, in, out, res))
                    return 0;

                bulk = AES_gcm_encrypt(in + res, out + res, len - res,
                                       ctx->gcm.key,
                                       ctx->gcm.Yi.c, ctx->gcm.Xi.u);

                ctx->gcm.len.u[1] += bulk;                                    // <-- wrap around happens here
                bulk += res;
            }
            if (CRYPTO_gcm128_encrypt_ctr32(&ctx->gcm, in + bulk, out + bulk,   // check for length here works with wrapped number and it passes wrong value
                                            len - bulk, ctx->ctr))
                return 0;
#else
            if (CRYPTO_gcm128_encrypt_ctr32(&ctx->gcm, in, out, len, ctx->ctr))
                return 0;
#endif /* AES_GCM_ASM */
        } else {                                                      
```

**Comment by slontis** [comment]
If it is broken, it has been like this for a long time... This code was copied from the old method in e_aes.c inside aes_gcm_cipher()

**Comment by amatyuko-intc** [comment]
After looking at it more, I think it is quite rare case to hit it in usual scenario. To face it you need to pass very big message to the function, so that its length is more than 64-bit accumulator can hold. It can be easier to reach if you already processed a message and the length of it is already close to the GCM input message bounds, but even in this case the last chunk should be at least 2^25 bit length.

**Comment by noahostle** [comment]
There is a rare but cryptographically meaningful impact to this bug, which I am adding here at the advice of openssl-security.

(H = GHASH Auth subkey, J0 = init counter block, S = AES_K(J0) = tag mask)

When an update crosses the NIST plaintext limit, the two overrun blocks expose S = AES_K(J0), where J0 = IV || 1. For IV = 0^96, the preceding wrapped counter block also exposes H = AES_K(0^128). I verified that these two values are sufficient to construct valid GCM tags without knowledge of the AES key via T=S+GHASH_H​(AAD,C).

I reproduced this on Ubuntu OpenSSL 3.0.2 and current upstream with the AES-NI backend selected.

Additionally, for any arbitrary 96-bit IV, the failed operation still leaks S, and if the application notices the failure, and subsequently retries under the same (K,V) and produces one valid full tag for a suitable message, H can be recovered algebraically from that tag via H = L^-1 · (T xor S) in, where L = [128]_64 || [0]_64 for a single zero AAD block and empty plaintext.

TLDR;

If an oracle can be invoked to cause the above bug, and is programmed to allow retries under the same IV on failure, an attacker can leak S and H and create valid tags for arbitrary AAD and ciphertext
